pkg/storage/rulestore/bucket_client.go

```diff
@@ -0,0 +1,311 @@
+package rulestore
+
+import (
+	"bytes"
+	"context"
+	"encoding/base64"
+	"fmt"
+	"io/ioutil"
+	"strings"
+
+	"github.com/go-kit/kit/log"
+	"github.com/go-kit/kit/log/level"
+	"github.com/gogo/protobuf/proto"
+	"github.com/thanos-io/thanos/pkg/objstore"
+	"golang.org/x/sync/errgroup"
+
+	"github.com/cortexproject/cortex/pkg/ruler/rules"
+	"github.com/cortexproject/cortex/pkg/tenant"
+)
+
+const (
+	delim      = "/"
+	rulePrefix = "rules" + delim
+)
+
+type BucketClient struct {
+	bucket          objstore.Bucket
+	loadConcurrency int
+
+	logger log.Logger
+}
+
+func NewBucketClient(bucket objstore.Bucket, loadConcurrency int, logger log.Logger) *BucketClient {
+	return &BucketClient{
+		bucket:          bucket,
+		loadConcurrency: loadConcurrency,
+		logger:          logger,
+	}
+}
+
+func (b *BucketClient) listNamespacesForUser(ctx context.Context, user string) ([]string, error) {
```

Comment:
I'm sorry. In all functions you will have to use `UserBucketClient` for every operation involving a specific user. We're leveraging on it to add extra functionalities and it's now required that we use it everywhere we leverage on the Thanos objstore client.

_We can discuss offline if you need more info about it._

Reply:
I transitioned all user operations to use a UserBucketClient. However, I had to make some additional changes to the UserBucketClient to allow for a global prefix, since all rules are currently stores on the path `rules/<user-id>`.